src/spikeinterface/core/node_pipeline.py

```diff
@@ -0,0 +1,698 @@
+            self.segment_slices.append(slice(i0, i1))
+
+    def get_trace_margin(self):
+        return 0
+
+    def get_dtype(self):
+        return base_peak_dtype
+
+    def compute(self, traces, start_frame, end_frame, segment_index, max_margin):
+        # get local peaks
+        sl = self.segment_slices[segment_index]
+        peaks_in_segment = self.peaks[sl]
+        i0 = np.searchsorted(peaks_in_segment["sample_index"], start_frame)
+        i1 = np.searchsorted(peaks_in_segment["sample_index"], end_frame)
+        local_peaks = peaks_in_segment[i0:i1]
+
+        # make sample index local to traces
+        local_peaks = local_peaks.copy()
+        local_peaks["sample_index"] -= start_frame - max_margin
+
+        return (local_peaks,)
+
+
+# this is not implemented yet this will be done in separted PR
+class SpikeRetriever(PeakSource):
+    """
+    This class is usefull to inject a sorting object in the node pipepline mechanisim.
+    It allows to compute some post processing with the same machinery used for sorting components.
+    This is a first step to totaly refactor:
+      * compute_spike_locations()
+      * compute_amplitude_scalings()
+      * compute_spike_amplitudes()
+      * compute_principal_components()
+
+
+    recording:
+
+    sorting:
+
+    channel_from_template: bool (default True)
+        If True then the channel_index is infered from template and extremum_channel_inds must be provided.
+        If False every spikes compute its own channel index given a radius around the template max channel.
+    extremum_channel_inds: dict of int
+        The extremum channel index dict given from template.
+    radius_um: float (default 50.)
+        The radius to find the real max channel.
+        Used only when channel_from_template=False
+    peak_sign: str (default "neg")
+        Peak sign to find the max channel.
+        Used only when channel_from_template=False
+    """
+
+    def __init__(
+        self, recording, sorting, channel_from_template=True, extremum_channel_inds=None, radius_um=50, peak_sign="neg"
+    ):
+        PipelineNode.__init__(self, recording, return_output=False)
+
+        self.channel_from_template = channel_from_template
+
+        assert extremum_channel_inds is not None, "SpikeRetriever need the dict extremum_channel_inds"
+
+        self.peaks = sorting_to_peak(sorting, extremum_channel_inds)
+
+        if not channel_from_template:
+            channel_distance = get_channel_distances(recording)
+            self.neighbours_mask = channel_distance < radius_um
+            self.peak_sign = peak_sign
+
+        # precompute segment slice
+        self.segment_slices = []
+        for segment_index in range(recording.get_num_segments()):
+            i0 = np.searchsorted(self.peaks["segment_index"], segment_index)
+            i1 = np.searchsorted(self.peaks["segment_index"], segment_index + 1)
+            self.segment_slices.append(slice(i0, i1))
```

Comment:
I think everywhere in SI where you are doing this, you should do
i0, i1 = np.searchsorted(self.peaks["segment_index"], [segment_index, segment_index + 1])

Should be faster

Reply:
Yes but this is absolutly everywhere.
A external PR for the entire code would be cool